Added MethodArgumentDefaultValueFixer

lmanzke opened the pull request:
Hey guys, glad to be doing my first PR for you :).

My colleague @dereuromark opened an issue for a new fixer suggestion a while ago.
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1462

Here is the PR implementing it. I took his solution and refactored it, please feel free to have a look.

I tried to make it as readable as possible. 

Maybe you have some suggestion on how finding the next occurrence of a Variable OR certain token might be done more conveniently. I thought I could use

```
$tokens->getNextTokenOfKind(T_VARIABLE) 
```

here, but it does not work that way ;).

Looking forward to your feedback!

Best regards,

Lucas

dereuromark commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file = null, $content)
```
`= null should be removed` - or you can run the fixer over itself. :) 
This should also be done for the other fixers in this repo if they contain this line.

lmanzke replied:
True ;)

dereuromark commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+    public function testFix($expected, $input = null)
+    {
+        $this->makeTest($expected, $input);
+    }
```
newline missing above

keradus commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+
+use Symfony\CS\Tests\Fixer\AbstractFixerTestBase;
+
+class MethodArgumentDefaultValueFixerTest extends AbstractFixerTestBase
```
final

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+     */
+    public function getPriority()
+    {
+        return -49;
```
why is that?

lmanzke replied:
Did not know what number to put there. Any suggestions?

gharlan replied:
If you do not know, try `0` (remove the method).

keradus replied:
you should think if the fixer you are creating need to run after or before any other existing fixer ?

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+ * @author Mark Scherer
+ * @author Lucas Manzke <lmanzke@outlook.com>
+ */
+class MethodArgumentDefaultValueFixer extends AbstractFixer
```
final

dereuromark replied:
why?

keradus replied:
All fixers are final at 2.x.
We cannot make them final on 1.x to not introduce BC breaks, but all new fixers should be final.

dereuromark replied:
That makes it impossible for them to be extended... :/

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+    {
+        $tokens = Tokens::fromCode($content);
+
+        $functionTokenIndices = array_keys($tokens->findGivenKind(T_FUNCTION));
```
don't use `findGivenKind` for that purpose.
That way you are iterating here twice.

keradus commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+     * @dataProvider provideExamples
+     *
+     * @param string $expected
+     * @param string $input
```
or null

keradus commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+EOT;
+
+        return array(
+            array('<?php public function bFunction($foo, $bar)', '<?php public function bFunction($foo = null, $bar)'),
```
please, write the cases like
```php
array(
    '<?php expected...',
    '<?php input...',
),
```

then it's more readable to see the difference

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+        $nextTokenIndex = $tokens->getNextTokenOfKind($index, $this->defaultValueTokens);
+        $nextToken = $tokens[$nextTokenIndex];
+
+        return $nextToken->getContent() === '=';
```
1. yoda with you should be
2. don't use getContent to check if the token is equals to sth. Use `equals` method

dereuromark replied:
You really should stop using Yoda, its a very bad practice - and instead use a sniffer/fixer to assert inline assignment is not possible (as we do).

keradus replied:
The code should be consistent. If whole project use it - it should be used here too.

Also, I don't think we have assignment in such a places anywhere.

dereuromark replied:
You can easily run a sniffer/fixer over it. I have written one. This way it is consistent then.

keradus replied:
And it is consistent.
Don't want to use yoda? it's cool. want to use yoda? it's cool. but don't mix.

dereuromark replied:
Yes, thats why I said you should stop doing it and fixing it (using an automated fixer/sniffer - at once).
Doesn't have to be today or tomorrow. But maybe in the near future.

keradus replied:
So far we do follow yoda condition and the PR should follow that too.
If you want to propose abandoning it - feel free to send separate PR for only that.

dereuromark replied:
I first have to publish the fixer for it, though.
Actually both NoYoda and NoInlineAssignment including sniffs.
So far they are still in development phase and not quite public-ready :)

lmanzke replied:
Well, it won't matter because I'll use equals anyway there. There won't be a comparison.

Is there a shortcut to check if the token is one of an array? Equals won't right cut it here.

lmanzke replied:
Saw you already suggest it ;) equalsAny is what I was looking for

keradus commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+            array('<?php public function dFunction($foo, $bar, $baz)', '<?php public function dFunction($foo = false, $bar, $baz)'),
+            array('<?php public function eFunction($foo, $bar, \SplFileInfo $baz, $x)', '<?php public function eFunction($foo = PHP_EOL, $bar, \SplFileInfo $baz = null, $x)'),
+            array('<?php public function eFunction($foo, $bar, \SplFileInfo $baz, $x = \'default\')', '<?php public function eFunction($foo, $bar = \'removedDefault\', \SplFileInfo $baz, $x = \'default\')'),
+            array($expected1, $input1),
```
just push them directly to array without that variables

dereuromark replied:
Doesnt make it even harder to read?
I find your current practice in the project of doing this inline strings quite unreadable.
It would be better to more often use elaborate <<< examples.

keradus replied:
When you have tons of different edge cases it's good to have many small cases instead few big ones.

keradus replied:
using heredoc/nowdoc is OK, but here catch them into variables are pointless.
that is why the names of them aren't great

keradus commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+                        public function fFunction($foo, $bar, \SplFileInfo $baz, $x = 'default');
+EOT;
+
+        return array(
```
missing a case like
```php
array(
    '<?php function foo($bar, $c) {}',
    '<?php function foo($bar /* a */ = /* b */ 1, $c) {}',
),
```

lmanzke replied:
Ok, will add one, but probably tomorrow. Am curious to see if it would work on that, don't see a problem yet though.

keradus replied:
I think it will pass, but it's great to have that test to protect the code to being corrupted during refactoring in the future

keradus commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,56 @@
+    {
+        $input1 = <<<'EOT'
+                    <?php
+                        public function eFunction($foo, $bar, \SplFileInfo $baz, $x = 'default';
```
All of the cases are invalid PHP code!

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+    {
+        $currentIndex = $nextVariableIndex;
+
+        while (!in_array($tokens[$currentIndex + 1]->getContent(), $this->argumentTerminatorTokens, true)) {
```
`equalsAny`

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,203 @@
+            return false;
+        }
+
+        while (!in_array($tokens[$nextRelevantTokenIndex]->getContent(), $this->functionDefinitionTerminatorTokens, true)) {
```
here and below - `equalsAny`

dereuromark commented:
Nice job!
I wonder if we should allso add an example with too less whitespace?

    public function hFunction($foo,$bar='removedValue',\SplFileInfo $baz,$x) {};

It should also remove the default value here properly without touching the rest of the whitespace.

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,76 @@
+        return array(
+            array(
+                '<?php public function bFunction($foo, $bar) {}',
+                '<?php public function bFunction($foo = null, $bar) {}',
```
please use valid php code

gharlan commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,201 @@
+    {
+        $examinedIndex = $tokens->getNextTokenOfKind($index, $this->argumentBoundaryTokens);
+        while (
+            $this->hasNonDefaultArgumentAfterIndex($tokens, $examinedIndex) &&
```
Couldn't you determine the last non-default argument index once per function definition?

lmanzke replied:
Yes, that would probably be an improvement. I'll change it a bit

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    /**
+     * @return int
```
inheritdoc

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,80 @@
+                ,
+                <<<'EOT'
+                    <?php
+                        function eFunction($foo, $bar, \SplFileInfo $baz, $x = 'default' {};
```
missing `)`

lmanzke replied:
This is actually intended, to check if in such an edge-case it does not do something strange.

gharlan replied:
we do not fix invalid files, so we do not have to test them.

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,193 @@
+    {
+        $variableIndices = array_keys($tokens->findGivenKind(T_VARIABLE));
+
+        $nextVariableIndex = $this->getFirstValueBiggerThan($variableIndices, $index);
```
`getNextTokenOfKind` should be faster than looking for all `T_VARIABLES` and getting first one.

lmanzke replied:
getNextTokenOfKind does not work for T_VARIABLE

lmanzke replied:
Its probably easier though just to iterate through the tokens until the first variable is encountered.

keradus replied:
Try this out:
```php
$test = $tokens->getNextTokenOfKind($index, [[T_VARIABLE]]);
```

second argument is an array of tokens you are looking for.
you are looking for token like `[T_VARIABLE, $foo, 123]`, but don't care about name or line, so simplify it into `[T_VARIABLE]`. And then you create an array with your token: `[[T_VARIABLE]]`

lmanzke replied:
Ok, didn't know this. I tried it with getNextTokenOfKind first and couldn't get it to work.

lmanzke replied:
Ok, see https://github.com/FriendsOfPHP/PHP-CS-Fixer/commit/b0f8a6aaecdea9f98504eeb0ec4f60afe7bde0ad . Got rid of the function to find the next variable or specific token all along.

keradus replied:
Great!
TBH I just don't want to have such a nice `getFirstValueBiggerThan` method :D

dereuromark commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    /**
+     * @param Tokens $tokens
```
If you do `@param Tokens|Token[] $tokens` you get typehinting on all `$tokens[$i]` etc.

lmanzke replied:
I get those with the present version, too. Don't you?

dereuromark replied:
Are you sure? Check the isGivenKind (is clickable?) in `$tokens[$i]->isGivenKind()` for example.

keradus replied:
I bet you are talking here about feature of PHPStorm, but looking on pure PHPDoc that notation would be wrong, because you cant pass `Token[]` to the method!

Typehinting of `$tokens[$nextRelevantIndex]` should be done thanks to https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.11/Symfony/CS/Tokenizer/Tokens.php#L23

dereuromark replied:
Didnt know that trick. Thx.

dereuromark commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,157 @@
+    public function fix(\SplFileInfo $file, $content)
+    {
+        $tokens = Tokens::fromCode($content);
+        for ($i = 0, $l = $tokens->count(); $i < $l; ++$i) {
```
Would need a protected/private method wrapping to have typehinting on the $tokens[$i]->...
But no big deal I guess.

lmanzke replied:
If you don't have it this is most likely a bug on your end, because my PHPStorm understands it the way it is :smile: . Meaning isGivenKind() can be autocompleted.

stof commented:
This fixer must not remove a ``null`` default value from a typehinted argument, even though it is followed by a required argument, as such default is about making the argument nullable, even though it cannot make it optional.
This case must be covered in the testsuite to ensure you don't break the code.

@keradus you should change the required status from appveyor as they splitted the status for pushes and PRs (as done by travis since a long time).

lmanzke commented:
@stof: I do not really see the difference between making an argument nullable and making it optional. Isn't a sign of poor design if nullable arguments have to be specified because they appear in the middle of the function definition?

I see your point, the fixer should not force you to refactor the code, but on the other side, this kind of refactoring should be done anyway.

lmanzke commented:
By the way, a general question on the tests: Why is it that expected and input must not be the same? Or is there a flag I have specify to point out that this behavior is intended?

I'd quite like to write some test-cases to test specifically that some constructions are **not** touched. However, that is disallowed, presumably as a precaution against tests that do not test anything by accident.

keradus commented:
```php
function foo(Bar $b = null, $c) {}

foo(new Bar(), 1); //OK
foo(null, 1); //OK too
```

if changed:
```php
function foo(Bar $b, $c) {}

foo(new Bar(), 1); //OK
foo(null, 1); //error
```

and that change the logic.

:+1: for @stof ;)

dereuromark commented:
@stof @keradus PHP makes all object attributes nullable either way. As such your examples are invalid as that is a default value, nothing to do with nullable.
To say that it can be nullable you should use the doc block `Object|null` instead.

keradus commented:
@lmanzke
take a look https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1495#issuecomment-155422233
also, going into `makeTest` method should also helps

dereuromark commented:
To clarify:

    function foo(Bar $b, $c) {}

Its perfectly valid to call `$this->foo(null, $c);` check it yourself

keradus commented:
@dereuromark 

```bash
ker@ad:/cli$ php --version
PHP 5.5.9-1ubuntu4.14 (cli) (built: Oct 28 2015 01:34:46) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies
    with Xdebug v2.2.3, Copyright (c) 2002-2013, by Derick Rethans
ker@ad:/cli$ 
ker@ad:/cli$ cat test.php 
<?php

class Bar{}
function fnc(Bar $b, $c) {}

fnc(null, 1);
ker@ad:/cli$ 
ker@ad:/cli$ php -f test.php 
PHP Catchable fatal error:  Argument 1 passed to fnc() must be an instance of Bar, null given, called in /cli/test.php on line 6 and defined in /cli/test.php on line 4
PHP Stack trace:
PHP   1. {main}() cli/test.php:0
PHP   2. fnc() /cli/test.php:6
ker@ad:/cli$ 
```

lmanzke commented:
@keradus Thanks for clearing that up. However, "go into the method" is a somewhat invalid argument. That should clearly be something immediately obvious from the docblock, which is not even present on the AbstractFixerTestBase->makeTest(). Maybe adding that would avoid confusion in the feature. (Also, having inheritdoc and then there is no parent docblock seems strange)

keradus commented:
> Maybe adding that would avoid confusion in the feature. 

great catch, feel free to attach it!

dereuromark commented:
Wooh, ok. I could have sworn it was different in some cases.

lmanzke commented:
I added the makeTest docblock in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1496

keradus commented:
ping ;)

lmanzke commented:
I added the case described by @stof, but also encountered something odd:
If I use getNextTokenOfKind('null'), it does not find it, as I half expected. Null seems to have the token id 310, which is, according to tokenizer.php, T_VARIABLE.

At runtime, however, T_VARIABLE is 312. Did I look up the wrong definition?
If yes, where do I find the right one? Don't like using "310" as magic token id at all.

gharlan commented:
The IDs are different in different php versions.

https://3v4l.org/psqa6

`null` is a `T_STRING` token.

gharlan commented:
> All tokens listed below are also defined as PHP constants. Their value is automatically generated based on PHP's underlying parser infrastructure. This means that the concrete value of a token may change between two PHP versions.

http://php.net/manual/en/tokens.php

lmanzke commented:
Good to know ;). The function token_name is also a good catch ;). Will change that then.
Still interesting that if it's a simple string that $token->equals('null') will not work.

keradus commented:
Don't rely on int value itself.

also, `getNextTokenOfKind` get the argument as array of possible tokens, not single token

lmanzke commented:
@keradus: I know  that, just was a bit lazy. Of course what I meant was getNextTokenOfKind(array('null')) won't work, because the underlying usage of equals will not catch it.

keradus commented:
```php
<?php

$a = null;

```

```php
$tokens = Tokens::fromCode($content);
$firstNull = $tokens->getNextTokenOfKind(0, [[T_STRING, 'null']]);
var_dump($firstNull, $tokens[$firstNull]->toJson());
```

```bash
int(6)                 
string(100) "{         
    "id": 310,         
    "name": "T_STRING",
    "content": "null", 
    "line": 3,         
    "isArray": true    
}"                     
```

lmanzke commented:
@keradus I start getting the impression that you do not read my comments / commits. I already adapted this. 
Still, even if your code works, it shows that the tokenizer is bad and unreliable. The getNextTokenOfKind function works for 'function', why not for 'null'? Why is it necessary to use the prototype array construction there?

Anyway, are there any things I should still adapt?

keradus commented:
@lmanzke you said that this looking for null won't work, I show it works. Sorry, I don't saw your comment that you have already figured it out.
One way or another - great that you manage it.

> Still, even if your code works, it shows that the tokenizer is bad and unreliable. The getNextTokenOfKind function works for 'function', why not for 'null'? Why is it necessary to use the prototype array construction there?

Because there is `T_FUNCTION` and there is not `T_NULL`. Please, send thanks to PHP authors. 


> Anyway, are there any things I should still adapt?

Is the PR request ready from your side? If yes we will start a review ;)

keradus commented:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493#issuecomment-155431991 is still to be done

dereuromark commented:
The test cases clearly show that this is not the case, that has been done already: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493/files#diff-5c2c894b7930f5a05b2dd0248e6c8d5bR82

keradus commented:
And still you have https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493/files#diff-5c2c894b7930f5a05b2dd0248e6c8d5bR37 .
You cannot assume that if a parameter is not typehinted as object it is not an object. It's safer to not delete `null` as default value.

If it will be removed for not-typehinted parameters then the fixer will be considered as risky (may change meaning of code!)

dereuromark commented:
> considered as risky

Maybe not a bad idea after all. In its original form it this fixer could be used to detect possible code smells early on.

But we could also fix it up to be non-risky and be done with it :)

keradus commented:
> In its original form it this fixer could be used to detect possible code smells early on.

Please remember that this is not mess detector but fixer.

lmanzke commented:
> If it will be removed for not-typehinted parameters then the fixer will be considered as risky (may 
> change meaning of code!)

I do not think that is true. That is the whole purpose if this fixer: If there is a `$a = null` somewhere in the function definition, the null does not make sense. It can be removed without changing the meaning of the code. The param is also nullable without the default value, but will always have to be passed if there are required arguments after the = null.

It is another case for typehinted arguments, though: Omitting the = null there would remove nullability and therefore change the behavior. So in this special case, if the argument is typehinted and default value is null, then it would change behavior, which is why it should leave it untouched.

However, it has to said that if something like this occurs, it is a code smell and should probably be refactored. But as this is not the purpose of the fixer, it leaves it as it is.

SpacePossum commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,87 @@
+            ),
+            array(
+                '<?php function foo ($bar, $c) {}',
+                '<?php function foo ($bar /* a */ = /* b */ 1, $c) {}',
```
Why are the comments removed?
What will this fixer do with something like:
```php
function foo(
  $a = 1, // test
  $b = 2, /* test*/
  $c = null, // abc
  $d 
) {}

dereuromark replied:
I agree, comments or non-related whitespace should not be removed (responsibility of other fixers if any).

keradus replied:
Some spaces should be removed, like:
`function foo($a =   1, $b) {}`
into:
`function foo($a, $b) {}`
is OK

`function foo($a    , $b) {}`
would be odd

dereuromark replied:
This is not a whitespace or argument spacing fixer though.

keradus replied:
If it remove piece of could it should take care not only about one token, but about place itself.

dereuromark replied:
When removing yes (every whitespace in between without comments). Otherwise it should skip.

keradus replied:
exactly, as in my example ;)

dereuromark replied:
Indeed, but @SpacePossum s example comments would not be touched by this, as there the extra stuff is after the value (and not between variable and value).

lmanzke replied:
I'll add test cases just to make sure.

SpacePossum replied:
ah you are right :)  thank for making sure

keradus replied:
I still do not like that the comments after variable are removed.
IMO both `/* a */` and `/* b */` comments should be there after removing default value

gharlan replied:
@keradus I agree

keradus replied:
in fact it looks like it is the last thing to do and good to merge then

gharlan replied:
imo this should be fixed: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493#discussion_r44389803

keradus replied:
yeah, good point

keradus commented:
@lmanzke , you are right, my bad.

lmanzke commented:
Testcase added and rebased it on top of current 1.11, too. I can also squash it a bit, but would do that at the end.

keradus commented:
(no need to squash yourself)

keradus commented:
@lmanzke it looks almost ready! Please, could you finish it off? I would be more than happy to merge it into 1.11

lmanzke commented:
So the open things are the comments that should not be removed and the "wrong" test case?

dereuromark commented:
@lmanzke Looks like it. Adding the closing braces sure makes sense as we only want to work with valid PHP anyway.

keradus commented:
exactly ;)

gharlan commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,167 @@
+
+        if (
+            $tokens[$nextMeaningfulTokenIndex]->equals(array(T_STRING, 'null')) &&
+            !$tokens[$prevMeaningfulTokenIndex]->equalsAny($this->argumentBoundaryTokens)
```
Please add a test case with a reference parameter:

```php
function foo(&$a = null, $b) {}
```

I'm not sure but I think it will fail because of the condition above.

gharlan replied:
And please add a test case with uppercased `NULL`. I'm not sure if it will pass.

```php
function foo(\SplFileInfo $file = NULL, $foo) {}
```
```

keradus replied:
And `Null` ;D

GrahamCampbell replied:
or, in otherwords, case insensitive :P

gharlan commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,167 @@
+        }
+
+        while (!$tokens[$currentIndex + 1]->equalsAny($this->argumentTerminatorTokens)) {
+            $tokens[++$currentIndex]->clear();
```
Please add this test case:

```php
function foo($foo = array(array(1)), $bar) {}
```

Should fail because `$this->argumentTerminatorTokens` contains `)`.

lmanzke commented:
@gharlan Added your test case and have adapted it to make it pass.
Currently having another one that is causing trouble though:

`public function a($a = array('a' => 'b', 'c' => 'd'), $b) {}`

lmanzke commented:
Wow, strange. Can somebody reproduce that this test case actually **breaks** the tokenizer?
`public function a($a = array('a' => 'b', 'c' => 'd'), $b) {}`

Tokens::fromCode returns one token for that string which is quite odd

lmanzke commented:
Seems to be a problem in the token_get_all function used by Tokens::fromCode.
My PHP Version: 
`PHP 5.6.15 (cli) (built: Nov 13 2015 14:32:10)
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Xdebug v2.3.3, Copyright (c) 2002-2015, by Derick Rethans`

lmanzke commented:
Ah see the error. Good old <?php is missing in front. Never mind ...

lmanzke commented:
Ok, I now added all test cases mentioned here and refactored the fixer to make them work. If there are any other test cases anybody can think of :wink:

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,136 @@
+            ),
+            array(
+                '<?php public function a($a, $b) {}',
+                '<?php public function a($a = array(\'a\' => \'b\', \'c\' => \'d\'), $b) {}',
```
invalid code (remove `public` or add a class around).

gharlan commented:
Missing this: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493#discussion_r45791923
(typehinted nullable argument with non-lowercase `NULL` (or `Null`).

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,136 @@
+            ),
+            array(
+                '<?php public function a($a, $b) {}',
+                '<?php public function a($a = [\'a\' => \'b\', \'c\' => \'d\'], $b) {}',
```
invalid code

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,136 @@
+            ),
+            array(
+                '<?php public function a($a, $b) {}',
+                '<?php public function a($a = NULL, $b) {}',
```
invalid code

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,136 @@
+            ),
+            array(
+                '<?php public function a($a, $b) {}',
+                '<?php public function a($a = Null, $b) {}',
```
invalid code

gharlan commented on Symfony/CS/Tests/Fixer/Contrib/MethodArgumentDefaultValueFixerTest.php:
```diff
@@ -0,0 +1,136 @@
+            ),
+            array(
+                '<?php public function a(&$a, $b) {}',
+                '<?php public function a(&$a = null, $b) {}',
```
invalid code

dereuromark replied:
I am pretty sure it was absolutely unnecessary to outline the same thing 5 times.

gharlan replied:
My comment: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493#discussion_r45812628

And the commit after that: https://github.com/lmanzke/PHP-CS-Fixer/commit/bcaee36cdd840ed16a89255bcbe389ac261cc05e

So I'm not sure. ;)

gharlan commented:
Oh, found another special case:

```php
function foo($a = 1, ...$b) {}
```

This is valid since PHP 5.6 and the default value must not be removed.
https://3v4l.org/EYN8Q

lmanzke commented:
Added a case for that and hope that all tests contain valid PHP code now.

lmanzke commented:
Ok, appveyor does not seem to know mb_convert_case (strange, exists since 4.3) and the T_ELLIPSIS constant is only available since 5.6 (makes sense). Will try to make sense of that tomorrow.

keradus commented:
As `mbstring` extension is not available by default in all envs and we * do not * rely on it.
Actually, we work to remove `ext-mbstring` from requirements in `compsoer.json` ;)

lmanzke commented:
Ok, tests pass again now as I put the ellipsis test cases into a separate method.

gharlan commented:
:+1: Thanks, very nice fixer!

lmanzke commented:
Seems I'm not quite there yet: 

`function a(\SplFileInfo &$a = null) {}` 

is valid php code (although senseless) and will probably fail. Verifying ... 

Would never have believed there are so many special cases! :wink:

lmanzke commented:
Actually, it passes. Added it to make sure though. 
Good to go.

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,236 @@
+    {
+        $currentIndex = $index - 1;
+
+        while ($tokens[$currentIndex]->isGivenKind(T_WHITESPACE)) {
```
it is not standard to have few `T_WHITESPACE` tokens one after another. document why it is possible here

lmanzke replied:
If that is the case, then the while can be removed, as it will only have one iteration ;)

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,236 @@
+    private function isEllipsis(Tokens $tokens, $variableIndex)
+    {
+        if (PHP_VERSION_ID < 50600) {
+            return false;
```
not sure about it...
if one run the fixer under 5.5 it doesn't mean the code that is fixing isn't written in 5.6

gharlan replied:
It is not possible to fix 5.6 codeunder 5.5.
Code is invalid, so the file would be skipped (after linting).

So Imo the solution above is ok.

keradus replied:
I remember we already doing it in few places.
But in that case, we should at least have a test for 5.6 which be fixed (we have it!) and a test for <5.6 with `...` which would not be changed.

gharlan replied:
> a test for <5.6 with ... which would not be changed.

It would be an invalid test (#1538). IMO we should avoid this.
The fixer would not fix the file under 5.5.

keradus replied:
And we should check if he skip it indeed.
Not all use the linting option.

lmanzke replied:
I thought there should be no invalid php code? For versions < 5.6, that would be the case if anything like ...$b occurs in a function definition. A test containing that would break that convention. The thing is it that the fixer would change PHP 5.5 code.

`<?php function a($a = 1, ...$b) {}`

would become

`<?php function a($a, ...b) {}` 

because the fixer would not 'understand' the ellipsis.

keradus replied:
Exactly. When I try to run the fixer on input code with ... using 5.5 interpreter, fixer should not remove the default value. We need to make prevention for that to happen. Because not all users use linting

lmanzke replied:
Ok that means there are invalid test cases, but at least it runs under 5.5. (and below)

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,236 @@
+        $lowerCasedNextContent = strtolower($tokens[$nextMeaningfulTokenIndex]->getContent());
+
+        return
+            $lowerCasedNextContent === 'null' &&
```
yoda with you should be

keradus commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,236 @@
+
+        return
+            $lowerCasedNextContent === 'null' &&
+            $tokens[$prevMeaningfulTokenIndex]->isGivenKind(T_STRING);
```
revert the condition. `T_STRING` could be check cheaply, then check that `null`

keradus commented on README.rst:
```diff
@@ -561,6 +561,12 @@ Choose from the list of available fixers:
                         Arrays should use the long
                         syntax.
 
+* **method_argument_default_value** [contrib]
```
Please move to Symfony

keradus commented:
:+1:

SpacePossum commented:
:+1:  thanks @lmanzke :)
before moving to Symfony level, I would like to know if @stof (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493#issuecomment-155431991) is happy with the implementation

keradus commented:
The fixer already follows his suggestion

gharlan commented on Symfony/CS/Fixer/Contrib/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,237 @@
+    private function isEllipsis(Tokens $tokens, $variableIndex)
+    {
+        if (PHP_VERSION_ID < 50600) {
+            return $tokens[$tokens->getPrevMeaningfulToken($variableIndex)]->equals('.');
```
@keradus do we really want to go this direction? Fixing weird token sequences (php 7 code tokenized with php 5 etc.)?

IMO if we do it here, we have to do it in many many other places.

Example: `$a <=> $b`
Here we have the spaceship operator (`T_SPACESHIP`), but php 5 tokenizer returns `<=` and `>` tokens, so we have to binary operators.
Result when fixing with php 5: `$a <= > $b`, which is invalid.

gharlan replied:
ping @keradus

keradus replied:
I'm ok if the PHP7 code won't be broken when fixing using PHP5 (no need to be fixed, just not broken)

gharlan replied:
`$a <=> $b` is "fixed" to `$a <= > $b` (using php5), which **is** broken.
I'm sure I could find other examples where the code is broken after fixing with wrong php version.

keradus replied:
One *bug* isn't a reason to introduce another one.

gharlan replied:
My opinion: linting should not be optional.
We should not try to handle php code of a newer version than the current running version.

But: accepted, we do not agree here. ;)

keradus replied:
I would be more than happy to have always linter mode on!
But - the performance loss is horrible :(

SpacePossum commented:
I don't think so, this test shows a default value `null` being removed because it is followed by a required argument https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493/files#diff-5c2c894b7930f5a05b2dd0248e6c8d5bR36

keradus commented:
check out https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1493#issuecomment-158511368 and following 4 comments

gharlan commented on Symfony/CS/Fixer/Symfony/MethodArgumentDefaultValueFixer.php:
```diff
@@ -0,0 +1,237 @@
+    {
+        $prevMeaningfulTokenIndex = $tokens->getPrevTokenOfKind($variableIndex, array(array(T_STRING), ',', '('));
+
+        if (!$tokens[$prevMeaningfulTokenIndex]->isGivenKind(T_STRING)) {
```
missing array and callable typehints.

keradus replied:
ping @lmanzke

lmanzke commented:
Didn't realize that PHP 5.3 doesn't know callable

GrahamCampbell commented:
callable is new in php 5.4, yeh

lmanzke commented:
Okay - callable is now also fixed in 5.3 , but without the constant

keradus commented:
Thank you @lmanzke.

SpacePossum commented:
thanks @lmanzke :)  :+1:

lmanzke commented:
np :)
